core/templates/expressions/expression-syntax-tree.service.ts

```diff
@@ -79,32 +108,32 @@ export class ExpressionSyntaxTreeService {
   constructor(private expressionParserService: ExpressionParserService) {}
 
   public getParamsUsedInExpression(expression: string): Array<string> {
-    let _findParams = (parseTree: Array<string>|string): Array<string> => {
-      let paramsFound = [];
-      if (parseTree instanceof Array) {
-        if (parseTree[0] === '#') {
-          paramsFound.push(parseTree[1]);
-        } else {
-          for (let i = 1; i < parseTree.length; i++) {
-            paramsFound = paramsFound.concat(_findParams(parseTree[i]));
-          }
-        }
-      }
+    let parsed = this.expressionParserService.parse(expression);
 
-      let uniqueParams = [];
-      for (let i = 0; i < paramsFound.length; i++) {
-        if (uniqueParams.indexOf(paramsFound[i]) === -1) {
-          uniqueParams.push(paramsFound[i]);
+    return this.findParams(parsed);
+  }
+
+  private findParams = (parseTree: Array<string>|string): Array<string> => {
+    let paramsFound = [];
+    if (parseTree instanceof Array) {
+      if (parseTree[0] === '#') {
+        paramsFound.push(parseTree[1]);
+      } else {
+        for (let i = 1; i < parseTree.length; i++) {
+          paramsFound = paramsFound.concat(this.findParams(parseTree[i]));
         }
       }
+    }
 
-      return uniqueParams.sort();
-    };
-
-    let parsed = this.expressionParserService.parse(expression);
+    let uniqueParams = [];
```

Comment:
Let's just use a `Set`?

```typescript
let paramsFound = new Set<string>();
paramsFound.add(value);
paramsFound = new Set([...paramsFound, ...this.findParams(parseTree[i]));
return paramsFound;

Reply:
@brianrodri I tried with `paramsFound` and I found it a bit hard when declaring `new Set<string>()` and then declaring inside else clause in the for-loop `new Set([...paramsFound, ...this.findParams(parse[i])]);

I tried with the `uniqueParams`:
```
let uniqueParams = new Set<string>();
    for (let i = 0; i < paramsFound.length; i++) {
      if (!uniqueParams.has(paramsFound[i])) {
        uniqueParams.add(paramsFound[i]);
      }
    }
    return Array.from(uniqueParams);
```

Reply:
I had the following in mind:

```typescript
  public getParamsUsedInExpression(expression: string): string[] {
    let parsedExpression = this.expressionParserService.parse(expression);
    return Array.from(this.findParams(parsedExpression)).sort();
  }

  private findParams = (parseTree: string|string[]): Set<string> => {
    let paramsFound = new Set<string>();
    if (parseTree instanceof Array) {
      if (parseTree[0] === '#') {
        paramsFound.add(parseTree[1]);
      } else {
        for (let i = 1; i < parseTree.length; i++) {
          paramsFound = new Set<string>([...this.findParams(parseTree[i]), ...paramsFound]);
        }
      }
      return paramsFound;
    }
  }
```

Reply:
If the unpacking is giving you trouble, we could also do:

```typescript
this.findParams(parseTree[i]).forEach(p => paramsFound.add(p));
```

Reply:
@brianrodri I tried that out but the linter tells that you need to activate --downLevelIteration. It works after adding it to tsconfig.js but I'm a bit worried if it doesn't add a lot more code after compiling it. I heard it can add multiple helper functions?

Don't have any troubles with unpacking, my end result:
```
public getParamsUsedInExpression(expression: string): string[] {
    let parsed = this.expressionParserService.parse(expression);

    return Array.from(this.findParams(parsed)).sort();
  }

  private findParams(parseTree: string[]|string): Set<string> {
    let paramsFound = new Set<string>();
    if (parseTree instanceof Array) {
      if (parseTree[0] === '#') {
        paramsFound.add(parseTree[1]);
      } else {
        for (let i = 1; i < parseTree.length; i++) {
          paramsFound = new Set<string>([...this.findParams(parseTree[i]),
            ...paramsFound]);
        }
      }
      return paramsFound;
    }

    let uniqueParams = new Set<string>();
    for (let i = 0; i < paramsFound.size; i++) {
      if (!uniqueParams.has(paramsFound[i])) {
        uniqueParams.add(paramsFound[i]);
      }
    }
    return uniqueParams;
  }
```

Reply:
I'm confused, on which line is that happening? I don't see any lines that looks like `paramsFound: Set<string>` in the code you pasted? Also, please remove the `uniqueParams` block, `Set` already guarantees unique values.

Reply:
This happens on this line of code:
```
paramsFound = new Set<string>([...this.findParams(parseTree[i]), ...paramsFound]);
```

Reply:
Got it, then let's just do the forEach approach (you can use `.values()` to bypass the type error):

```typescript
this.findParams(parseTree[i]).values().forEach(p => paramsFound.add(p));
```

Reply:
@brianrodri Ok, I used this approach without `uniqueParams` block and got 1 test failing:
```
HeadlessChrome 80.0.3987 (Mac OS X 10.14.6) Expression evaluator service should get params used in expressions FAILED
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
```
Doesn't say a lot to me. I tried to console.log the length and items in it but nothing suspecious.

Reply:
Hmm, are you sure you don't see any extra debug information? I took a look at that code ([link](https://github.com/oppia/oppia/blob/develop/core/templates/expressions/expression-evaluator.service.spec.ts#L61)), and it looks like it tries to print out a lot of info about what failed and why:
```
      var recordFailure = function(params, exception) {
        console.error('input           : ' + expression);
        console.error('parsed          : ' + parsedJson);
        console.error('expected        : ' + JSON.stringify(expectedParams));
        if (params !== undefined) {
          console.error('evaluated       : ' + params);
        } else {
          console.error('exception       : ' + exception);
        }
        failed = true;
      };
```

Reply:
@brianrodri Now I understand:
It's not giving the expected parsed result at is should be when looking in `expression-evaluator.service.spec.ts`:
```
HeadlessChrome 80.0.3987 (Mac OS X 10.14.6): Executed 1148 of 1972 SUCCESS (0 secs / 31.221 secs)
ERROR: 'input           : log(9, 3)'
HeadlessChrome 80.0.3987 (Mac OS X 10.14.6): Executed 1148 of 1972 SUCCESS (0 secs / 31.221 secs)
ERROR: 'parsed          : ["log",9,3]'
HeadlessChrome 80.0.3987 (Mac OS X 10.14.6): Executed 1148 of 1972 SUCCESS (0 secs / 31.221 secs)
ERROR: 'expected        : []'
HeadlessChrome 80.0.3987 (Mac OS X 10.14.6): Executed 1148 of 1972 SUCCESS (0 secs / 31.221 secs)
ERROR: 'exception       : TypeError: Cannot read property 'forEach' of undefined'
```

Reply:
Hmm weird, OK let's do this: Please file an issue to eventually change this function to use `Set`, and assign it to me. Then leave a `TODO(#ISSUE_NUM):` describing the same thing and we can save it for later -- it isn't important enough to block you any more.

Reply:
@brianrodri I agree, what about the code itself? Should I wait until the issue gets resolved/merged and then fetch the updated `oppia:develop`?

Reply:
Nope, I'm fine with merging this as is! Thanks for all the hard work :)